### ISSUE TITLE
Support for leader election

### ISSUE THREAD

**Issue by dkellenb** [etcd-io/jetcd#408]
It would be great if jetcd would support leader election. I checked i would be able to implement this based on `jetcd` based on following shell script: https://gist.github.com/ueokande/01ec5cdd6fa31f9bd481dc05e7d42c50

What I miss is how I can implement the requests with jetcd:

`curl -sSL -XPUT "${LEADER_KEY}?prevValue=${ID}&ttl=${TTL}&value=${ID}"` and `curl -sSL -XPUT "${LEADER_KEY}?prevExist=false&ttl=${TTL}&value=${ID}"`, I miss following parameters on `PutOption`:
- `prevValue`
- `ttl`
- `prevExists` (or is this the `withPrevKV` ?)

Probably it would be even better to support [elect](https://github.com/etcd-io/etcd/blob/master/etcdctl/README.md#elect-options-election-name-proposal)

**Comment by lburgazzoli** [comment]
It should be implemented using protobuf api instead of rest, I think there is a protobuf spec on etcd side

**Comment by pengfeiwang-cn** [comment]
does jetcd plan to support election?

**Comment by lburgazzoli** [comment]
yes it is but help would be more than happy :)

**Comment by lukasz-antoniak** [comment]
Hi @lburgazzoli, I have done a small PoC here: https://github.com/lukasz-antoniak/jetcd/commit/8e4c74e858eb872a0a585c77e1a942995f279378. This is my first contribution to the project, but if you think that code looks more or less acceptable, I will work on it further.

**Comment by lburgazzoli** [comment]
It looks good however I see there is a proto definition for leader election, like https://github.com/etcd-io/etcd/blob/master/etcdserver/api/v3election/v3electionpb/v3election.proto so wonder if we should base the leader election client to such service as it requires less maintenance on the client side

**Comment by lukasz-antoniak** [comment]
Argh, not sure why I missed that. I have been focusing on Golang client implementation. Updated version here: https://github.com/lukasz-antoniak/jetcd/commit/46dd80d7cb7903416765e896f190700efe7b55c0 (as a plain driver, mostly proxying calls to server). I would still like to provide an interface that automatically handles registration and revocation of lease. Maybe it is a good idea to introduce to simplified interface in _jetcd-extensions_ module?

**Comment by lburgazzoli** [comment]
The code looks good so feel free to open a PR (same for the second question)

**Comment by lukasz-antoniak** [comment]
Thank you for having quick look. Submitted pull request https://github.com/etcd-io/jetcd/pull/724. I will work on simplified interface in the _extensions_ module once base functionality is merged.

**Comment by lburgazzoli** [comment]
@lukasz-antoniak merged

**Comment by lukasz-antoniak** [comment]
While working on extended client, I have realized about wrong classification of two exceptions. `not a leader` indicates that we are trying to do operation allowed only for a leader. `no leader` means that etcd has no leader for certain election. Sorry for that. You merged previous pull request when I was just about to fix the issue. Opened new pull request: https://github.com/etcd-io/jetcd/pull/726.

**Comment by lburgazzoli** [comment]
@lukasz-antoniak can we close this issue or do you have any additional work to do ?

**Comment by lukasz-antoniak** [comment]
@lburgazzoli, yes I think so. I have been trying to implement easier version, but I think it had too little value - only maintenance of lease, which users may want still to do on their own.
